aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -380,11 +384,35 @@ module aptos_framework::account {
     }
 
     /// A resource account is used to manage resources independent of an account managed by a user.
-    public fun create_resource_account(source: &signer, seed: vector<u8>): (signer, SignerCapability) {
-        let addr = create_resource_address(&signer::address_of(source), seed);
-        let signer = create_account_unchecked(copy addr);
-        let signer_cap = SignerCapability { account: addr };
-        (signer, signer_cap)
+    /// In Aptos a resource account is created based upon the sha3 256 of the source's address and additional seed data.
+    /// A resource account can only be created once, this is designated by setting the
+    /// `Account::signer_capbility_offer::for` to the address of the resource account. While an entity may call
+    /// `create_account` to attempt to claim an account ahead of the creation of a resource account, if found Aptos will
+    /// transition ownership of the account over to the resource account. This is done by validating that the account has
+    /// yet to execute any transactions and that the `Account::signer_capbility_offer::for` is none. The probability of a
+    /// collision where someone has legitimately produced a private key that maps to a resource account address is less
+    /// than `(1/2)^(256).
+    public fun create_resource_account(source: &signer, seed: vector<u8>): (signer, SignerCapability) acquires Account {
+        let resource_addr = create_resource_address(&signer::address_of(source), seed);
+        let resource = if (exists_at(resource_addr)) {
+            let account = borrow_global<Account>(resource_addr);
+            assert!(
```

Comment:
These asserts apply in the general case too right? Wonder if it's worth refactoring as.. 

```
let resource = if (exists_at(resource_addr)) {
  let account = borrow_global<Account>(resource_addr);
  create_signer(resource_addr)
} else {
   create_account_unchecked(resource_addr)
};

assert!(...)

let account = borrow_global_mut<Account>(resource_addr);
account.signer_capability_offer.for = option::some(resource_addr);
let signer_cap = SignerCapability { account: resource_addr };
(resource, signer_cap)
```

This has the advantage of making it easier to reason that `account.signer_capability_offer.for = option::some(resource_addr);` will never overwrite an existing option, guaranteeing that only one `SignerCapability` will ever get created. It has the disadvantage of creating the `signer` object prior to doing any checks. Either way works, but another potential option

Reply:
Creating the signer prior to any checks is not a big problem if the tx would revert when later assertions fail. Either code makes sense to me. What David has right now makes sense because if the account doesn't exist, there's no signer cap offer to check anyway. Both have the same exact effect I think so either works 🤷